File: backend/documents/services/placeholder_normalization.py
from __future__ import annotations

import re

from .variable_keys import CANONICAL_ASSET_LOGO, CANONICAL_ASSET_SIGNATURE


CURRENT_DATE_KEYS = {
    "date",
    "validity_date",
    "document.current_date",
    "document.validity_date",
}

_ASSET_ALIASES = {
    "assets.logo": CANONICAL_ASSET_LOGO,
    "asset_logo": CANONICAL_ASSET_LOGO,
    "assets_logo": CANONICAL_ASSET_LOGO,
    "logo": CANONICAL_ASSET_LOGO,
    "company_logo": CANONICAL_ASSET_LOGO,
    "[logo]": CANONICAL_ASSET_LOGO,
    "__asset_logo__": CANONICAL_ASSET_LOGO,
    "company.logo": CANONICAL_ASSET_LOGO,
    "assets.signature": CANONICAL_ASSET_SIGNATURE,
    "asset_signature": CANONICAL_ASSET_SIGNATURE,
    "assets_signature": CANONICAL_ASSET_SIGNATURE,
    "signature": CANONICAL_ASSET_SIGNATURE,
    "company_signature": CANONICAL_ASSET_SIGNATURE,
    "[signature]": CANONICAL_ASSET_SIGNATURE,
    "__asset_signature__": CANONICAL_ASSET_SIGNATURE,
    "company.signature": CANONICAL_ASSET_SIGNATURE,
}


def extract_placeholder_segments(raw: str) -> list[str]:
    token = (raw or "").strip()
    if not token:
        return []

    if token.startswith("{{") and token.endswith("}}"):
        token = token[2:-2].strip()

    if token.startswith("__ASSET_") or token.startswith("["):
        token = token.split(":", 1)[0].strip()

    parts = [part.strip() for part in re.split(r"[|,]", token) if part.strip()]
    return parts


def canonicalize_placeholder_key(raw: str) -> str:
    segments = extract_placeholder_segments(raw)
    if not segments:
        return ""

    lowered = re.sub(r"\s+", "", segments[0]).lower()
    if not lowered:
        return ""

    return _ASSET_ALIASES.get(lowered, lowered)


def placeholder_has_modifier(raw: str, modifier: str) -> bool:
    normalized_modifier = re.sub(r"\s+", "", (modifier or "")).lower()
    if not normalized_modifier:
        return False
    return any(
        re.sub(r"\s+", "", part).lower() == normalized_modifier
        for part in extract_placeholder_segments(raw)[1:]
    )


def is_current_date_placeholder(*, raw: str | None = None, canonical_key: str | None = None) -> bool:
    key = canonicalize_placeholder_key(raw or "") if raw else re.sub(r"\s+", "", canonical_key or "").lower()
    if not key:
        return False

    if key in CURRENT_DATE_KEYS:
        return True

    if raw and placeholder_has_modifier(raw, "date"):
        return True

    return key.endswith("_date")

File: src/lib/document-workflow/placeholder-normalization.ts
const ASSET_ALIASES: Record<string, string> = {
	'assets.logo': 'assets.logo',
	asset_logo: 'assets.logo',
	assets_logo: 'assets.logo',
	logo: 'assets.logo',
	company_logo: 'assets.logo',
	'[logo]': 'assets.logo',
	__asset_logo__: 'assets.logo',
	'company.logo': 'assets.logo',
	'assets.signature': 'assets.signature',
	asset_signature: 'assets.signature',
	assets_signature: 'assets.signature',
	signature: 'assets.signature',
	company_signature: 'assets.signature',
	'[signature]': 'assets.signature',
	__asset_signature__: 'assets.signature',
	'company.signature': 'assets.signature',
};

const CURRENT_DATE_KEYS = new Set([
	'date',
	'validity_date',
	'document.current_date',
	'document.validity_date',
]);

export function extractPlaceholderSegments(raw: string): string[] {
	let token = (raw || '').trim();
	if (!token) return [];

	if (token.startsWith('{{') && token.endsWith('}}')) {
		token = token.slice(2, -2).trim();
	}

	if (token.includes('{') || token.includes('}')) {
		return [];
	}

	if (token.startsWith('__ASSET_') || token.startsWith('[')) {
		token = token.split(':', 1)[0]?.trim() ?? token;
	}

	return token
		.split(/[|,]/)
		.map(part => part.trim())
		.filter(Boolean);
}

export function canonicalizePlaceholderKey(raw: string): string {
	const segments = extractPlaceholderSegments(raw);
	if (segments.length === 0) return '';

	const normalized = segments[0].replace(/\s+/g, '').toLowerCase();
	if (!normalized) return '';

	return ASSET_ALIASES[normalized] ?? normalized;
}

export function placeholderHasModifier(raw: string, modifier: string): boolean {
	const normalizedModifier = modifier.replace(/\s+/g, '').toLowerCase();
	return extractPlaceholderSegments(raw)
		.slice(1)
		.some(
			segment =>
				segment.replace(/\s+/g, '').toLowerCase() === normalizedModifier,
		);
}

export function isDatePlaceholder(raw: string): boolean {
	const canonical = canonicalizePlaceholderKey(raw);
	if (!canonical) return false;
	return (
		CURRENT_DATE_KEYS.has(canonical) ||
		placeholderHasModifier(raw, 'date') ||
		canonical.endsWith('_date')
	);
}

export function formatCurrentDateInputValue(date = new Date()): string {
	const year = date.getFullYear();
	const month = String(date.getMonth() + 1).padStart(2, '0');
	const day = String(date.getDate()).padStart(2, '0');
	return `${day}.${month}.${year}`;
}

export function describePlaceholderSource(
	rawSource: string | null | undefined,
): string {
	switch ((rawSource || '').toUpperCase()) {
		case 'IMPORTED':
			return 'Auto-filled';
		case 'AI':
			return 'AI draft';
		case 'COMPOSED':
			return 'Composed';
		case 'MANUAL':
			return 'Saved';
		default:
			return 'Pending';
	}
}
